### PR TITLE
Add detailed developer information about adding pair styles and use new born/gauss style as example

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This expands the documentation about adding new styles to LAMMPS with a detailed description about adding  pair styles to augment the existing example for writing fixes.

**Related Issue(s)**

N/A

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

The pair style is unusual, since its application is for liquid mercury at different temperatures and the parameterization is tunable. So it will also give an example to use ``Pair::extract()`` and fix adapt for a simulation with a temperature ramp.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] One or more example input decks are included

### CODE REVIEW

File: doc/src/Developer_write_pair.rst [hunk 1]
```diff
@@ -0,0 +1,1196 @@
+Writing new pair styles
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Pair styles are at the core of most simulations with LAMMPS, since they
+are used to compute the forces (plus energy and virial contributions, if
+needed) on atoms for pairs of atoms within a given cutoff.  This is
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: pairs or small clusters of atoms within a given cutoff

File: doc/src/Developer_write_pair.rst [hunk 2]
```diff
@@ -0,0 +1,1196 @@
...
+other methods are optional and have default implementations in the base
+class (most of which do nothing), but they may need to be overridden
+depending on the requirements of the model.
+
+Case 1: a pairwise additive model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this section, we will describe the procedure of adding a simple pair
+style to LAMMPS: an empirical model that can be used to model liquid
+mercury.
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: also say which pair style this is within LAMMPS and where the file is, so users can look at the actual pair file as they read this section ?

File: doc/src/Developer_write_pair.rst [hunk 3]
```diff
@@ -0,0 +1,1196 @@
...
+   // clang-format off
+   PairStyle(born/gauss,PairBornGauss);
+   // clang-format on
+   #else
+
+   /* the definition of the PairBornGauss class (see below) is inserted here */
+
+   #endif
+
+This second segment of the header file will be included by the ``Force``
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: confused by "segment" - I think you simply mean the PairStyle() line above ?  if so, say that

File: doc/src/Developer_write_pair.rst [hunk 4]
```diff
@@ -0,0 +1,1196 @@
...
+energy with :doc:`pair_write <pair_write>`).  This is a repetition of
+the force kernel in the ``compute()`` function, but only for a single
+pair of atoms, where the (squared) distance is provided as a parameter
+(so it may not even be an existing distance between two specific atoms).
+The energy is returned as the return value of the function and the force
+as the `fforce` reference.  Note, that this is, similar to how *fpair*
+is used in the ``compute()`` function, the magnitude of the force along
+the vector between the two atoms *divided* by the distance.
+
+The ``single()`` function is optional.  The member variable
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: Explain that any true pairwise style should define this method.  Manybody methods do not.

File: doc/src/Developer_write_pair.rst [hunk 5]
```diff
@@ -0,0 +1,1196 @@
...
+   }
+
+
+Reading and writing of restart files
+""""""""""""""""""""""""""""""""""""
+
+Support for writing and reading binary restart files is provided by the
+following four functions.  Writing is only done by MPI processor rank 0.
+The output of global (not related to atom types) settings is delegated
+to the ``write_restart_settings()`` function.  Implementing the
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: Explain that if the pair style only has a handful of per-type params, it should define these methods.
If there are large tabulation files (e.g. EAM) or manybody it should not.

File: doc/src/Developer_write_pair.rst [hunk 6]
```diff
@@ -0,0 +1,1196 @@
...
+     MPI_Bcast(&cut_global, 1, MPI_DOUBLE, 0, world);
+     MPI_Bcast(&offset_flag, 1, MPI_INT, 0, world);
+     MPI_Bcast(&mix_flag, 1, MPI_INT, 0, world);
+   }
+
+Writing coefficients to data files
+""""""""""""""""""""""""""""""""""
+
+The ``write_data()`` and ``write_data_all()`` functions are optional and
+write out the current state of the :doc:`pair_coeff
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst:848]: Also explain this is only relevant for true pairwise styles.  Manybody pair styles don't write to data files.

File: doc/src/Developer_write_pair.rst [hunk 7]
```diff
@@ -0,0 +1,1196 @@
...
+         fprintf(fp, "%d %d %g %g %g %g %g %g\n", i, j, biga0[i][j], alpha[i][j], biga1[i][j],
+                 beta[i][j], r0[i][j], cut[i][j]);
+   }
+
+
+Give access to internal data
+""""""""""""""""""""""""""""
+
+The purpose of the ``extract()`` function is to allow access to internal data
+of the pair style to other parts of LAMMPS.  One application is to use
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: by other parts of LAMMPS

File: doc/src/Developer_write_pair.rst [hunk 8]
```diff
@@ -0,0 +1,1196 @@
...
+
+Case 2: a many-body potential
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Since there is a detailed description of the purpose and general layout
+of a pair style in the previous case, we will focus on where the
+implementation of a typical many-body potential *differs* from a
+pair-wise additive potential.  We will use the implementation of the
+Tersoff potential as :doc:`pair_style tersoff <pair_tersoff>` as an
+example.
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: tell the user where the tersoff source files are

File: doc/src/Developer_write_pair.rst [hunk 9]
```diff
@@ -0,0 +1,1196 @@
...
+     }
+   [...]
+
+
+Reading potential parameters
+""""""""""""""""""""""""""""
+
+For the Tersoff potential, the parameters are listed in a file and associated
+with triples for elements.  Thus, the ``coeff()`` function has to do three
+tasks, each of which is delegated to a function:
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: explain the * * requirement of the pair coeff command for many manybody potentials.
Which means that coeff() is only called once, unlike for pairwise styles

File: doc/src/Developer_write_pair.rst [hunk 10]
```diff
@@ -0,0 +1,1196 @@
...
+For some models, the interactions between atoms depends on properties of
+their environment which have to be computed *before* the the forces can
+be computed.  Since LAMMPS is designed to run in parallel using a
+:doc:`domain decomposition strategy <Developer_par_part>`, not all
+information of the atoms may be directly available and thus
+communication steps may be need to collect data from ghost atoms of
+neighboring subdomains or send data to ghost atoms for application
+during the pairwise computation.  For this we will look at how the
+embedding term of the :doc:`embedded atom potential EAM <pair_eam>` is
+implemented in LAMMPS.
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst]: point user to EAM source files

File: doc/src/Developer_write_pair.rst [hunk 11]
```diff
@@ -0,0 +1,1196 @@
...
+   if (atom->nmax > nmax) {
+     memory->destroy(rho);
+     memory->destroy(fp);
+     memory->destroy(numforce);
+     nmax = atom->nmax;
+     memory->create(rho,nmax,"pair:rho");
+     memory->create(fp,nmax,"pair:fp");
+     memory->create(numforce,nmax,"pair:numforce");
+   }
+
```
Comment (sjplimp) [inline doc/src/Developer_write_pair.rst:1165]: Before the following detailed sections, add one sentence to define in general what LAMMPS means by forward comm and one to define reverse comm.  E.g. forward = copying owned per-atom data to ghost atoms, reverse = summing ghost per-atom data to owned atoms.

**Comment by akohlmey** [comment]
@sjplimp this should now address all your suggestions and a few additional improvements.

**Review by sjplimp — APPROVED** [review]

**Pull request merged**